Sources/NIO/ByteBuffer-views.swift

```diff
@@ -16,48 +16,161 @@
+    }
+
+    public mutating func withUnsafeMutableBytes<R>(range: Range<Index>, _ body: (UnsafeMutableRawBufferPointer) throws -> R) rethrows -> R {
+        precondition(self._range.lowerBound <= range.lowerBound && self._range.upperBound >= range.upperBound,
+                     "range is out of bounds")
+
+        return try self._buffer.withVeryUnsafeMutableBytes { ptr in
+            try body(UnsafeMutableRawBufferPointer(start: ptr.baseAddress!.advanced(by: range.lowerBound),
+                                                   count: range.count))
         }
     }
 
     public var startIndex: Index {
-        return self.range.lowerBound
+        return self._range.lowerBound
     }
 
     public var endIndex: Index {
-        return self.range.upperBound
+        return self._range.upperBound
     }
 
     public func index(after i: Index) -> Index {
         return i + 1
     }
 
     public subscript(position: Index) -> UInt8 {
-        guard position >= self.range.lowerBound && position < self.range.upperBound else {
-            preconditionFailure("index \(position) out of range")
+        get {
+            guard position >= self._range.lowerBound && position < self._range.upperBound else {
+                preconditionFailure("index \(position) out of range")
+            }
+            return self._buffer.getInteger(at: position)! // range check above
+        }
+        set {
+            guard position >= self._range.lowerBound && position < self._range.upperBound else {
+                preconditionFailure("index \(position) out of range")
+            }
+            self._buffer.setInteger(newValue, at: position)
         }
-        return self.buffer.getInteger(at: position)! // range check above
     }
 
     public subscript(range: Range<Index>) -> ByteBufferView {
-        return ByteBufferView(buffer: self.buffer, range: range)
+        get {
+            return ByteBufferView(buffer: self._buffer, range: range)
+        }
+        set {
+            self.replaceSubrange(range, with: newValue)
+        }
+    }
+
+    @inlinable
+    public mutating func replaceSubrange<C: Collection>(_ subrange: Range<Index>, with newElements: C) where ByteBufferView.Element == C.Element {
+        precondition(subrange.lowerBound >= self.startIndex && subrange.upperBound <= self.endIndex,
+                     "subrange out of bounds")
+
+        if newElements.count <= subrange.count {
+            // Copy the bytes over into the subrange.
+            self._copyBytes(to: subrange, from: newElements)
+
+            // `newElements` is shorter than subrange, so we now have unwanted bytes between the
+            // copied bytes and the end of the subrange. Let's move all of the bytes after the
+            // subrange taking care to update the buffer's writer index as well.
+            if newElements.count < subrange.count {
+                let oldWriterIndex = self._buffer.writerIndex
```

Comment:
instead of doing all of this inline, could we create a `ByteBuffer` function

    public func moveBytes(at atIndex: Int, to toIndex: Int, length: Int) {
        ...
    }

or something similar